### PR TITLE
TransactionSignature MetaData support

### PR DESCRIPTION
**Pull request by kchalkias**
A `SignableData` is now signed, which is a wrapper over txId (merkle root) and an extra metadata object added from the signer.
A `TransactionSignature` object is then sent instead of the previously used `DigitalSignature.WithKey` with the main difference that the metadata should be included, so the receiver can reconstruct the `SignableData` to verify the signature.

transactionsignature_with_meta - page 1

Currently `platformVersion` and schemeNumberID (signing cipher suite) are the sole metadata used. The class does not allow inheritance but in the future we may allow CorDapp developers to extend it, so they can add other valuable data, such as a user's timestamp or other app-specific information.

Note that `DigitalSignature.WithKey` hasn't been removed, because is is still used in other than tx signatures like `WireNodeRegistration` and notary errors.

### CODE REVIEW

**Comment by kchalkias** [comment]
The mentioned flaky test seems to be independent on this work and can't be reproduced in Linux or Mac, only on Windows. FYI it is the `detect double spend 2 faulty` test in `BFTNotaryServiceTests`. `detect double spend 1 faulty` also fails some times in my machine.

**Comment by rnicoll** [comment]
Can we pull the general cleanup (i.e. removing `IllegalArgumentException` from `@Throws`, correcting case in comments) into its own PR to try isolating this into functional changes, please? Makes it a lot easier to review/discuss usefully.

**Comment by kchalkias** [comment]
Thanks Ross, will do first think tomorrow. I just took the opportunity for these tiny changes as it was initially required on the Metadata-related refactored methods, then replicated this to the rest of similar signing/verify methods. But you are right, it would be easier for the review process to remove them from this PR.

File: core/src/main/kotlin/net/corda/core/crypto/MerkleRootWithMeta.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * Using a [MerkleRootWithMeta] object a signer can add extra information on the transaction signature.
+ * It actually works as a wrapper that contains the Merkle root along with extra transactionMetaData, such as DLT's platform version.
+ *
+ * @param merkleRoot the merkle root of the transaction.
+ * @param transactionSignatureMeta meta data required.
+ */
+@CordaSerializable
+class MerkleRootWithMeta(val merkleRoot: SecureHash,
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/MerkleRootWithMeta.kt]: As the root itself is reconstructed by the recipient of the signature, perhaps this should fix the signature scheme in use by using `SecureHash.SHA256` here. We may end up with multiple merkle roots in future as we migrate between hash functions e.g. if we wish to introduce a parallel subset-sum based tree.

File: core/src/main/kotlin/net/corda/core/crypto/MerkleRootWithMeta.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+                         val transactionSignatureMeta: TransactionSignatureMeta) {
+
+    fun bytes() = this.serialize().bytes
+
+    override fun equals(other: Any?): Boolean {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/MerkleRootWithMeta.kt]: Why not use a data class for this?

File: core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt [hunk 3]
```diff
@@ -1,22 +1,47 @@
...
 
 /**
- * A wrapper around a digital signature accompanied with metadata, see [MetaData.Full] and [DigitalSignature].
- * The signature protocol works as follows: s = sign(MetaData.hashBytes).
+ * A wrapper around a digital signature accompanied with metadata.
+ * This is similar to [DigitalSignature.WithKey], but targeted to DLT transaction signatures.
  */
-open class TransactionSignature(val signatureData: ByteArray, val metaData: MetaData) : DigitalSignature(signatureData) {
+@CordaSerializable
+open class TransactionSignature(bytes: ByteArray, val by: PublicKey, val transactionSignatureMeta: TransactionSignatureMeta): DigitalSignature(bytes) {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt]: I'd rather not have it be open for now. We can open it later, once we've had time to think this through in detail, but not close it again. As we lack compelling use cases for app-specific stuff here ... let's leave it closed.

File: core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt [hunk 4]
```diff
@@ -1,22 +1,47 @@
...
      */
-    @Throws(InvalidKeyException::class, SignatureException::class, IllegalArgumentException::class)
-    fun verify(): Boolean = Crypto.doVerify(metaData.publicKey, signatureData, metaData.bytes())
+    @Throws(InvalidKeyException::class, SignatureException::class)
+    fun verify(merkleRoot: SecureHash) = Crypto.doVerify(merkleRoot, this)
+
+    @Throws(InvalidKeyException::class, SignatureException::class)
+    fun isValid(merkleRoot: SecureHash) = Crypto.isValid(merkleRoot, this)
+
+    override fun equals(other: Any?): Boolean {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt]: Making it no longer open means we can use a data class here too.

File: core/src/main/kotlin/net/corda/core/crypto/TransactionSignatureMeta.kt [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * TransactionSignatureMeta is required to add extra meta-data to a transaction's signature.
+ * It currently supports platformVersion only, but it can be extended to support a universal digital
+ * signature model enabling partial signatures and attaching extra information, such as a user's timestamp or other
+ * application-specific fields.
+ *
+ * @param platformVersion current DLT version.
+ */
+@CordaSerializable
+open class TransactionSignatureMeta(val platformVersion: Int) {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/TransactionSignatureMeta.kt]: Ditto, let's close it and use `data class`. Also do we need an algorithm specifier here?

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt [hunk 6]
```diff
@@ -134,12 +136,17 @@ interface ServiceHub : ServicesForResolution {
...
-    fun signInitialTransaction(builder: TransactionBuilder, publicKey: PublicKey): SignedTransaction {
-        val sig = keyManagementService.sign(builder.toWireTransaction().id.bytes, publicKey)
+    fun signInitialTransaction(builder: TransactionBuilder, publicKey: PublicKey, transactionSignatureMeta: TransactionSignatureMeta): SignedTransaction {
+        val merkleRootWithMeta = MerkleRootWithMeta(builder.toWireTransaction().id, transactionSignatureMeta)
+        val sig = keyManagementService.sign(merkleRootWithMeta, publicKey)
         builder.addSignatureUnchecked(sig)
         return builder.toSignedTransaction(false)
     }
 
+    /** The default [TransactionSignatureMeta] is used*/
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: Can you reverse the api docs here so the one the user is likely to use (this one) has the full docs, and the other uses a cross-reference to it with the longer description about when you'd want to customise them ... which is probably never, now I think about it, so why are we adding public API here?

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt [hunk 7]
```diff
@@ -173,10 +180,13 @@ interface ServiceHub : ServicesForResolution {
...
      * If the [PublicKey] is actually a [net.corda.core.crypto.CompositeKey] the first leaf key found locally will be used
      * for signing.
-     * @return The [DigitalSignature.WithKey] generated by signing with the internally held [java.security.PrivateKey].
+     * @return The [TransactionSignature] generated by signing with the internally held [java.security.PrivateKey].
      */
-    fun createSignature(signedTransaction: SignedTransaction, publicKey: PublicKey): DigitalSignature.WithKey {
-        return keyManagementService.sign(signedTransaction.id.bytes, publicKey)
+    fun createSignature(signedTransaction: SignedTransaction, publicKey: PublicKey) = createSignature(signedTransaction, publicKey, TransactionSignatureMeta(myInfo.platformVersion))
+
+    fun createSignature(signedTransaction: SignedTransaction, publicKey: PublicKey, transactionSignatureMeta: TransactionSignatureMeta): TransactionSignature {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: Same issue here except this new function doesn't even _have_ public API docs.

File: core/src/main/kotlin/net/corda/core/node/services/KeyManagementService.kt [hunk 8]
```diff
@@ -53,4 +55,7 @@ interface KeyManagementService {
      */
     @Suspendable
     fun sign(bytes: ByteArray, publicKey: PublicKey): DigitalSignature.WithKey
+
+    @Suspendable
+    fun sign(merkleRootWithMeta: MerkleRootWithMeta, publicKey: PublicKey): TransactionSignature
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/node/services/KeyManagementService.kt]: Needs API doc.

File: core/src/main/kotlin/net/corda/core/node/services/VaultService.kt [hunk 9]
```diff
@@ -106,7 +106,7 @@ class Vault<out T : ContractState>(val states: Iterable<StateAndRef<T>>) {
      *  results will be empty)
      */
     @CordaSerializable
-    data class Page<out T : ContractState>(val states: List<StateAndRef<T>>,
+    data class Page<T : ContractState>(val states: List<StateAndRef<T>>,
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/node/services/VaultService.kt]: Why did you remove the variance here?

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 10]
```diff
@@ -153,18 +153,18 @@ open class TransactionBuilder(
...
-        addSignatureUnchecked(key.sign(data.bytes))
+        val merkleRootWithMeta = MerkleRootWithMeta(toWireTransaction().id, TransactionSignatureMeta(1)) // A dummy platformVersion.
+        addSignatureUnchecked(key.sign(merkleRootWithMeta))
         return this
     }
 
     /** Adds the signature directly to the transaction, without checking it for validity. */
     @Deprecated("Use ServiceHub.signInitialTransaction() instead.")
-    fun addSignatureUnchecked(sig: DigitalSignature.WithKey): TransactionBuilder {
+    fun addSignatureUnchecked(sig: TransactionSignature): TransactionBuilder {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: You might as well just delete the deprecated methods rather than change their signature, which would require the dev to port from one deprecated api to another!

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt [hunk 11]
```diff
@@ -153,18 +153,18 @@ open class TransactionBuilder(
 
     /** The signatures that have been collected so far - might be incomplete! */
     @Deprecated("Signatures should be gathered on a SignedTransaction instead.")
-    protected val currentSigs = arrayListOf<DigitalSignature.WithKey>()
+    protected val currentSigs = arrayListOf<TransactionSignature>()
 
     @Deprecated("Use ServiceHub.signInitialTransaction() instead.")
     fun signWith(key: KeyPair): TransactionBuilder {
-        val data = toWireTransaction().id
-        addSignatureUnchecked(key.sign(data.bytes))
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: delete

**Comment by kchalkias** [comment]
Thanks Mike,
After careful consideration, we may need to have `MerkleRootWithMeta` and `TransactionSignature` store the **serialised** `TransactionSignatureMeta`, so the notary is agnostic on the metadata type used - no need to deserialise or whitelist custom metadata classes that CorDapps may use in the future.

The actual question is: Does the notary need to know the Metadata type used?
If not, then when a CorDapp uses (or will use if we make it open in the future) custom app-specific Metadata, there is no need to update the notary nodes (by implementing and whitelisting the custom Metadata classes).
If we leave it as it now and only support not dynamic, closed Metadata, then any future change on the metadata object will require a change (and potential restarts) on the notary nodes.

**Comment by mikehearn** [comment]
> so the notary is agnostic on the metadata type used - no need to deserialise or whitelist custom metadata classes that CorDapps may use in the future.

Bear in mind that the current situation with serialisation is temporary. Rick's work is nearing completion and after that is done, we can deserialise types we do not already know about safely. Classes will be synthesised on the fly to hold them so they can be reflected over, and as these classes are generated client-side we know them to be safe.

So we shouldn't be structuring types to try and control what gets deserialised when. We can just let services like notaries go ahead and deserialise whatever they want. At most it wastes a bit of work creating objects and filling them out that the receiving code will then just ignore.

File: core/src/main/kotlin/net/corda/core/crypto/MerkleRootWithMeta.kt (hunk 1 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/MerkleRootWithMeta.kt]: Should this apply to the `id: SecureHash` used in Transaction classes as well? If changed on this class, I have to do it there as well or always check `txId as SecureHash.SHA256`.

File: core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt (hunk 3 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt]: We use a `NullSignature` class in DummyKeys.kt that extends it.

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt (hunk 10 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: I was expecting to do on another PR. They are already used already by various tests, thus I have to update them first before deleting the deprecated.

File: core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt (hunk 4 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt]: `TransactionSignature` extends `DigitalSignature(bytes)`, thus is it possible to convert it to a data class?

File: core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt (hunk 10 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: I think Ross is working on removing deprecated methods. It turns out that it should be in another PR anyway as they are still heavily used. Are you ok with that?

Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/transactions/TransactionBuilder.kt]: OK

File: core/src/main/kotlin/net/corda/core/crypto/Crypto.kt [hunk 12]
```diff
@@ -401,23 +402,23 @@ object Crypto {
     }
 
     /**
-     * Generic way to sign [MetaData] objects with a [PrivateKey].
-     * [MetaData] is a wrapper over the transaction's Merkle root in order to attach extra information, such as a timestamp or partial and blind signature indicators.
+     * Generic way to sign [SignableData] objects with a [PrivateKey].
+     * [SignableData] is a wrapper over the transaction's id (merkle root) in order to attach extra information, such as a timestamp or partial and blind signature indicators.
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: Merkle is named after Ralph Merkle, and therefore should have a capital 'M'.

Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: Done for Merkle root, tree and leaf

File: core/src/main/kotlin/net/corda/core/crypto/SignatureMetadata.kt [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+/**
+ * SignatureMeta is required to add extra meta-data to a transaction's signature.
+ * It currently supports platformVersion only, but it can be extended to support a universal digital
+ * signature model enabling partial signatures and attaching extra information, such as a user's timestamp or other
+ * application-specific fields.
+ *
+ * @param platformVersion current DLT version.
+ */
+@CordaSerializable
+data class SignatureMetadata(val platformVersion: Int, val schemeNumberID: Int)
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/SignatureMetadata.kt]: What's "schemeNumberID" and why is it not documented?

File: core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt [hunk 14]
```diff
@@ -1,22 +1,57 @@
...
+     *
+     * @throws InvalidKeyException if the key to verify the signature with is not valid (i.e. wrong key type for the
+     * signature).
+     * @throws SignatureException if the signature is invalid (i.e. damaged).
+     * @return whether the signature is correct for this key.
+     */
+    @Throws(InvalidKeyException::class, SignatureException::class)
+    fun isValid(txId: SecureHash) = Crypto.isValid(txId, this)
+
+    override fun equals(other: Any?): Boolean {
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt]: Can't this be a data class now?

File: core/src/main/kotlin/net/corda/core/node/services/KeyManagementService.kt [hunk 15]
```diff
@@ -53,4 +55,17 @@ interface KeyManagementService {
...
\ No newline at end of file
+
+    /**
+     * Using the provided signing [PublicKey] internally looks up the matching [PrivateKey] and signs the [SignableData].
+     * @param signableData a wrapper over transaction id (Merkle root) and signature metadata.
+     * @param publicKey The [PublicKey] partner to an internally held [PrivateKey], either derived from the node's primary identity,
+     * or previously generated via the [freshKey] method.
+     * If the [PublicKey] is actually a [CompositeKey] the first leaf signing key hosted by the node is used.
+     * @throws IllegalArgumentException if the input key is not a member of [keys].
+     * TODO: A full [KeyManagementService] implementation needs to record activity to the Audit Service and to limit signing to
```
Comment (mikehearn) [inline core/src/main/kotlin/net/corda/core/node/services/KeyManagementService.kt]: This TODO should be in the implementation classes, right?

File: core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt (hunk 14 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/TransactionSignature.kt]: Well, there is an issue where `TransactionSignature` extends `DigitalSignature(bytes)`, which extends OpaqueBytes(bytes). If I make `TransactionSignature` a data class by overriding `bytes`, then the `init` function on `OpaqueBytes` fails with `NullPointerException`.

File: core/src/main/kotlin/net/corda/core/node/ServiceHub.kt [hunk 16]
```diff
@@ -132,25 +135,30 @@ interface ServiceHub : ServicesForResolution {
...
-        builder.addSignatureUnchecked(sig)
-        return builder.toSignedTransaction(false)
-    }
+    fun signInitialTransaction(builder: TransactionBuilder, publicKey: PublicKey) =
+            signInitialTransaction(builder, publicKey, SignatureMetadata(myInfo.platformVersion, Crypto.findSignatureScheme(publicKey).schemeNumberID))
 
     /**
      * Helper method to construct an initial partially signed transaction from a TransactionBuilder
-     * using the default identity key contained in the node.
+     * using the default identity key contained in the node. The legal Indentity key is used to sign.
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/node/ServiceHub.kt]: "legal Indentity key" should be "legal identity key"
